### ISSUE TITLE
Kotlinx collection immutable support

### ISSUE THREAD

**Issue by beigirad** [square/moshi#1739]
We're replacing [Kotlinx collection](https://github.com/Kotlin/kotlinx.collections.immutable) with normal collections in our project. But we can't use this types in json classes that Moshi requires.

There's some alternative JsonAdapters and Factories  (like #449 and some other gist instances) but they're not as well as built-in adapters and hasn't full support like `List` and `Map`

**Comment by ZacSweers** [comment]
Is there a question here?

**Comment by beigirad** [comment]
@ZacSweers 
No. I chose feature-request template. I have request to implement it as an artifact to extends Moshi features.

If you're agree with it, I can contribute for it.

**Comment by ZacSweers** [comment]
There's a similar request in moshix (moshi extensions), I think it'd make more sense there. The PR stalled, but perhaps you could offer to take it over - https://github.com/ZacSweers/MoshiX/pull/478
